### ISSUE TITLE
Lint for redundant keyword_init on structs

### ISSUE THREAD

**Issue by aldhsu** [rubocop/rubocop#13501]
## Is your feature request related to a problem? Please describe.

Ruby 3.2 enables keyword_init on structs without specifying it. https://www.ruby-lang.org/en/news/2022/12/25/ruby-3-2-0-released/

```rb
# before 3.2
foo = Struct.new(:bar, keyword_init: true)

# equivalent after 3.2
foo = Struct.new(:bar)
```

## Describe the solution you'd like

A lint for redundant keyword_init: true on structs.

**Comment by koic** [comment]
This change introduces a behavioral incompatibility with the `Struct.keyword_init?` method, so it may be worth considering whether "redundant" is an appropriate designation.

```ruby
Struct.new(:bar, keyword_init: true).keyword_init? # => true
Struct.new(:bar).keyword_init? # => nil
```

**Comment by aldhsu** [comment]
Good catch @koic!
Do you think that behaviour makes sense in the `keyword_init?` method? 

The default has changed. Consumers relying on `!Struct.new(:bar).keyword_init?` will not be aware of the change. I'm not sure what the patterns usually are for `?` methods in Ruby. However, I think in my experience, the expectation is that `nil` will be treated as falsy rather than truth-y for `?` methods.

**Comment by koic** [comment]
Yeah, RuboCop is used widely across applications, libraries, and frameworks, and user programs can include unexpected cases. Rather than focusing on whether it is a predicate method, it is important to consider the potentials for altering existing behavior, which differs from layout changes.

**Comment by aldhsu** [comment]
OK that makes sense.

> consider the potentials for altering existing behavior, which differs from layout changes.

Would it make sense to have the cop disabled by default with the safety warning that `keyword_init?` doesn't respond the same when using keyword_init explicitly?

**Comment by koic** [comment]
Technically, for Ruby 3.2+, it seems possible to pending it by default with `SafeAutoCorrect: false`.

**Comment by Earlopain** [comment]
This is an easy one to implement and I think it makes sense. I'd like to keep this open for now.

Only for >= 3.2 makes sense. Here's the feature request for ruby for adding `keyword_init?` method: https://bugs.ruby-lang.org/issues/18008. Making it unafe seems ok but I can't imagine it mattering much.

**Comment by Earlopain** [comment]
Ref https://github.com/rubocop/rubocop/pull/13893#discussion_r1966495970

Should this be closed? I learned that `keyword_init: true` does more than the name suggests, and also blocks initialization with positional arguments entirely.

Is this something to care about? I think, most add `keyword_init: true` to stay compatible with older ruby versions.

**Comment by javierjulio** [comment]
I'm not sure if this argument is really redundant, unfortunately, since it depends on how the Struct is initialized. While I think that warrants a bug on the app code, I still wanted to share this after enabling this cop. I was surprised to find that removing it on Ruby 3.4.x didn't raise an error if the Struct still happened to be initialized with a hash instead of keyword arguments but the latter was assumed.

```
Ruby 3.4.8p72
>> WithKeywordInit = Struct.new(:a, :b, :c, keyword_init: true)
>> WithKeywordInit.new(a: 1, b: 2, c: 3)
=> #<struct WithKeywordInit a=1, b=2, c=3>
>> WithKeywordInit.new({ a: 1, b: 2, c: 3 })
=> #<struct WithKeywordInit a=1, b=2, c=3>
>> NoKeywordInit = Struct.new(:a, :b, :c)
>> NoKeywordInit.new({ a: 1, b: 2, c: 3 })
=> #<struct NoKeywordInit a={a: 1, b: 2, c: 3}, b=nil, c=nil>
```

Note that without the `keyword_init: true` argument, if initialized with a hash, the first Struct key gets the hash set as its value but when using the argument, either a hash or keyword arguments, will set the Struct as expected.

I figured I'd share this to see if it warrants an issue report or not. Thank you.

**Comment by oleksii-leonov** [comment]
As @javierjulio mentioned, `keyword_init: true` changes behaviour in several ways, so it's not correct to mark it as "redundant".

One more important notice is that `keyword_init: true` allows only keyword arguments and raises an exception if someone tries to use positional arguments. This is often desired behaviour to enforce one coding style and better code robustness (e.g. it's safe to change the order of arguments without breaking existing code).

**Comment by oleksii-leonov** [comment]
It may be good to rename it from `Style/RedundantStructKeywordInit` to `Style/StructKeywordInit` and add a configuration parameter `EnforcedStyle` with two options:

- `allow_only_keyword_init_true` (enforce usage of `keyword_init: true`)
- `disallow` (prohibit usage of `keywored_init: true`)

**Comment by bbatsov** [comment]
@rubocop/rubocop-core I wonder if we should disable this cop by default, given the discussion above.

**Comment by koic** [comment]
Yeah, disabling this by default seems to reduce user confusion. I've opened PR #15063.

**Comment by strika** [comment]
@oleksii-leonov already explained that:

> One more important notice is that `keyword_init: true` allows only keyword arguments and raises an exception if someone tries to use positional arguments. This is often desired behaviour to enforce one coding style and better code robustness (e.g. it's safe to change the order of arguments without breaking existing code).

But, I feel we should also consider the current vs old Ruby documentation.

[Ruby 3.2 Release Notes](https://www.ruby-lang.org/en/news/2022/12/25/ruby-3-2-0-released/) stated that:

> A Struct class can also be initialized with keyword arguments without `keyword_init: true` on `Struct.new`.

Which suggests that `keyword_init:` is really redundant.

However, [Ruby 4.0 Documentation](https://docs.ruby-lang.org/en/4.0/Struct.html) states:

> By default, the arguments for initializing an instance of the new subclass can be both positional and keyword arguments.

> Optional keyword argument `keyword_init:` allows to force only one type of arguments to be accepted:

So, it clearly states that `keyword_init:` is not redundant, but that it alters the behavior of the class. It suggests that this is its main purpose. Because of that, I find this cop to be confusing.

It's useful to have in mind that Standard [enables](https://github.com/standardrb/standard/blob/v1.55.0/config/base.yml#L1775-L1776) this cop by default. Which will confuse many people. This may be an issue with Standard, but this cop itself is confusing.
